### PR TITLE
Slippage Correction

### PR DESCRIPTION
**Pull request by bh2smith**
This Generates all the same data that has been manually put together here in [this spreadsheet](https://docs.google.com/spreadsheets/d/18Q4_W-0vz1F3YXNLkue46JWgvND9pQZPpgMOXBvVlro/edit#gid=119754741) but was easier to combine the different months. We generate and commit a full csv file with the total owed also [here as a formatted spreadsheet](https://docs.google.com/spreadsheets/d/1IRQWX4mRLK3cUJhu5VIiHKO3RXKE7jVP50jzEDdRApo/edit?usp=sharing).

Attached is also a ZIP file of all the Query Results.

[SlippageCSVs.zip](https://github.com/cowprotocol/solver-rewards/files/11937929/SlippageCSVs.zip) (Updated to include May)

Total ETH Owed (by us red - to us green) January to April -- does not yet include May.
Screenshot 2023-05-25 at 09 55 27

Note there is one very strange ETH price inconsistency in a calculation commented [here](https://docs.google.com/spreadsheets/d/18Q4_W-0vz1F3YXNLkue46JWgvND9pQZPpgMOXBvVlro/edit?disco=AAAAxpSK5yc)

### CODE REVIEW

**Comment by bh2smith** [comment]
I have updated the linked PR with a CSV containing the largest differences (per transaction).
There are only 50 records (out of 100 THOUSAND) which differ in absolute value by more than $100.
Nearly all of the differences are cases where the one slippage calculation was ZERO and the other was wrong.
[Here is a spreadsheet](https://docs.google.com/spreadsheets/d/1LJsrMF6t-BvYHfFuQwt7gImfldlOMoiLGhUFW-mAiV8/edit#gid=973137515) with the 50 results differing by over 100.
The largest one was [this](https://dune.com/queries/2421375?StartTime_d83555=2023-01-01+00%3A00%3A00&EndTime_d83555=2023-04-30+00%3A00%3A00&CTE_NAME_e15077=incoming_and_outgoing_with_internalized_imbalances&TxHash_t6c1ea=0x18c3ab88737a1a863aac28264b05f359325b1c5267074cfb84c7bb0aef9de649) and it is clear (to me) that the new calculation is the correct one.
I will still spend some time today examining a few cases, but I think we are generally good to go! This new slippage calculation is a MAJOR improvement in both accuracy and computational efficiency!



If anyone has a chance to take a look at this, we could already do next week's payout with the new thang.

File: src/scripts/slippage_correction.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        old_df = pd.read_csv(
+            dune.get_result_csv(OLD_SLIPPAGE_EXECUTION_IDS[month]).data
+        )
+        new_df = pd.read_csv(
+            dune.get_result_csv(NEW_SLIPPAGE_EXECUTION_IDS[month]).data
+        )
+
+        # Take out the trash
+        del old_df["batchwise_breakdown"]
+        del new_df["batchwise_breakdown"]
```
Comment (gentrexha) [inline src/scripts/slippage_correction.py]: Are you trying to get rid of those specific columns here? Dropping the columns here using `.drop` is a lot more understandable than using `del`. Ideally, you would not read this column at all and could do:

```python
old_df = pd.read_csv(
            dune.get_result_csv(OLD_SLIPPAGE_EXECUTION_IDS[month]).data, usecols=["cols", "I_want", ...]
        )
```

File: src/scripts/slippage_correction.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        combined_df = join_coalesce_fillna(old_df, new_df)
+
+        combined_df["eth_diff_wei"] = (
+            combined_df["eth_slippage_wei_old"] - combined_df["eth_slippage_wei_new"]
+        )
+        # These don't actually matter, but we can have them if we want.
+        combined_df["us_diff"] = (
+            combined_df["usd_value_old"] - combined_df["usd_value_new"]
+        )
+        # combined_df["eth_diff"] = combined_df["eth_diff_wei"] / 1e18
```
Comment (gentrexha) [inline src/scripts/slippage_correction.py]: Forgotten comment?

File: src/scripts/slippage_correction.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    """
+    merged = pd.merge(df_1, df_2, on=join_col, how="outer", suffixes=list(suffixes))
+    try:
+        # coalesce solver names
+        col = "solver_name"
+        merged[col] = merged[f"{col}{suffixes[0]}"].combine_first(
+            merged[f"{col}{suffixes[1]}"]
+        )
+        del merged[f"{col}{suffixes[0]}"]
+        del merged[f"{col}{suffixes[1]}"]
```
Comment (gentrexha) [inline src/scripts/slippage_correction.py]: Same comment as on other `del` instance.

File: src/scripts/slippage_correction.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    per_tx_df = join_coalesce_fillna(
+        df_1=pd.read_csv(dune.get_result_csv(PER_TX_SLIPPAGE["old"]).data),
+        df_2=pd.read_csv(dune.get_result_csv(PER_TX_SLIPPAGE["new"]).data),
+        join_col="tx_hash",
+    )
+    per_tx_df["diff_us"] = per_tx_df["usd_value_new"] - per_tx_df["usd_value_old"]
+
+    # Everything differing by 100 USD in absolute value.
+    reduced = per_tx_df[abs(per_tx_df["diff_us"]) > 100]
+    reduced.to_csv("./biggest_diff.csv", index=False)
```
Comment (gentrexha) [inline src/scripts/slippage_correction.py]: I think saving these to a data directory would be a good idea. What do you think?

File: src/scripts/slippage_correction.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    for month, diff_df in month_diffs.items():
+        agg_result = join_coalesce_fillna(
+            agg_result, diff_df, suffixes=(str(month - 1), str(month))
+        )
+
+    agg_result["total_wei"] = sum(agg_result[f"eth_diff_wei{i}"] for i in months)
+    agg_result["total_us"] = sum(agg_result[f"us_diff{i}"] for i in months)
+    agg_result["total_eth"] = agg_result["total_wei"] / 1e18
+
+    agg_result.to_csv("./slippage_correction.csv", index=False)
```
Comment (gentrexha) [inline src/scripts/slippage_correction.py]: Same as reduced.

**Review by gentrexha — COMMENTED** [review]
Left some nitpick comments about the code. 

Not really sure how I'm supposed to validate the correctness of the slippage correction just from the information here. Would be interested for you to walk me through some of your findings.

File: src/scripts/slippage_correction.py (hunk 4 above)
Comment (bh2smith) [reply src/scripts/slippage_correction.py]: yea we could do that.

**Comment by bh2smith** [comment]
> Not really sure how I'm supposed to validate the correctness of the slippage correction just from the information here. Would be interested for you to walk me through some of your findings.

You can look at any of the transactions in the [largest diff file](https://docs.google.com/spreadsheets/d/1LJsrMF6t-BvYHfFuQwt7gImfldlOMoiLGhUFW-mAiV8/edit#gid=973137515) - check etherscan and compare the results of old slippage query with new one on a per transaction basis.

Some queries of interest are:

- [Batch Token Imbalance](https://dune.com/queries/1955401?TxHash=0x5ad2dbdb9adcee92e7a3514538492ded217f4945437da46f2c1917fee3a61ca6&sidebar=version-history&CTE_NAME_e15077=final_results&TxHash_t6c1ea=0xF1032C1C883CFE2DADB6CF03D7019ABA08071AB561CAA3440945F98C81A29F1C)

and probably best is `final_token_balance_sheet` CTE in the new Period Solver Slippage vs the old one:

https://github.com/cowprotocol/solver-rewards/pull/259#issuecomment-1564170268

Period Solver Slippage: 
- NEW - https://dune.com/queries/2421375 
- OLD - https://dune.com/queries/2259597

I would set the start and end date to January 1 - May 26 and then paste whatever `TxHash` you are interested in for each.

File: src/scripts/slippage_correction.py (hunk 3 above)
Comment (bh2smith) [reply src/scripts/slippage_correction.py]: This is necessary, the result of join does not necessarily have a solver_name in both. we need to coalesce and then delete. If we only did use_cols, we could potentially wind up with a null solver name.

**Comment by gentrexha** [comment]
Thanks. I'll check this out.

**Comment by bh2smith** [comment]
Among the list of 50 transactions (all of which we should expect were internalized -- to explain the diff). we found that 48 / 50 were internalized and the other two (VERY WEIRD) batches were PERFECT COWS (although they were both super similar and containing two trades with the same owners. They were like self manufactured cows.

```sql
select distinct tx_hash from internalized_imbalances where tx_hash in HASHES_FROM_DIFF_LIST
```

These are the weird ones (NOTE that they look almost identical but are different). 

1. https://etherscan.io/tx/0xd6801c9418c723466cec9f79b1b2719e1b1ee80b072d08ae45cac86bd639c5bc
2. https://etherscan.io/tx/0xd60a4f64b42ec05732b3326402c018619ea38a48274e46360365a75fcacd3f78


There is other weird stuff here.... like how the transfers look like there were two trades, but there was only one trade. I think something is unusual about [this token](https://etherscan.io/token/0xa8b919680258d369114910511cc87595aec0be6d) -- maybe it emits two transfer events on each transfer... not sure.

 
It appears that the buffer trade approximation assumed they were buffer trades (being perfect cows) and determined no slippage although there was slippage and there was no internalization.

**Comment by gentrexha** [comment]
> There is other weird stuff here.... like how the transfers look like there were two trades, but there was only one trade. I think something is unusual about [this token](https://etherscan.io/token/0xa8b919680258d369114910511cc87595aec0be6d) -- maybe it emits two transfer events on each transfer... not sure.

I think this seems to be the case since looking at these two in our explorer ([1](https://explorer.cow.fi/tx/0xd6801c9418c723466cec9f79b1b2719e1b1ee80b072d08ae45cac86bd639c5bc?tab=orders), [2](https://explorer.cow.fi/tx/0xd60a4f64b42ec05732b3326402c018619ea38a48274e46360365a75fcacd3f78?tab=orders)) there seems to be only one order showing up. On the other hand, if you look at the graph, we can see both coins being sent and received.

**Review by gentrexha — APPROVED** [review]

File: src/scripts/slippage_correction.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+# TODO - Include Whatever Part of May.
+months = [1, 2, 3, 4]
+
+OLD_SLIPPAGE_EXECUTION_IDS = {
+    1: "01H155A02SH9J58K79KW6581DN",
+    2: "01H155MD0SSYDENMWE6Y6DRE37",
+    3: "01H15A3JJFH8E9ZSG0GVPN3963",
+    4: "01H15ADMN4Q9X63DAW6FVSY1W7",
+    # 5: ""
```
Comment (bh2smith) [inline src/scripts/slippage_correction.py]:
```suggestion
    5: "01H4DSXSK7FZD14BQ8FKFKV43X",
```

File: src/scripts/slippage_correction.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
+"""Basic script combining a bunch of Dune CSV results to correct Slippage calculation."""
+import os
+
+import pandas as pd
+from dune_client.client import DuneClient
+from dotenv import load_dotenv
+
+# TODO - Include Whatever Part of May.
+months = [1, 2, 3, 4]
```
Comment (bh2smith) [inline src/scripts/slippage_correction.py]:
```suggestion
months = [1, 2, 3, 4, 5]
```

File: src/scripts/slippage_correction.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    4: "01H15ADMN4Q9X63DAW6FVSY1W7",
+    # 5: ""
+}
+
+NEW_SLIPPAGE_EXECUTION_IDS = {
+    1: "01H15924T2A8ZMSDSJXAZTDW7K",
+    2: "01H159ED65QE417NWAGFVAC6HB",
+    3: "01H171P567EEDSKRYPMP419WR2",
+    4: "01H1721SX8XS4HCDVWFHXJQPFA",
+    # 5: ""
```
Comment (bh2smith) [inline src/scripts/slippage_correction.py]:
```suggestion
    5: "01H4DSKSGV0FT5CMJVBZ7NQEJQ",
```

**Review by bh2smith — COMMENTED** [review]
Adding query results for May (only to May 29 -- the last day we used the old payouts).

**Comment by fhenneke** [comment]
I think the slippage correction has been applied a long time ago. So I am closing this.

**Pull request closed without merging**